### PR TITLE
WT-10957 Remove LFENCE and SFENCE from WiredTiger along with weaker memory model macros.

### PR DESCRIPTION
**Pull request by luke-pearson**
RIP WT_ORDERED_READ_WEAK_MEMORDER you won't be missed.

### CODE REVIEW

File: src/include/msvc.h [hunk 1]
```diff
@@ -114,23 +114,7 @@ WT_PAUSE(void)
 static inline void
 WT_READ_BARRIER(void)
```
Comment (ajmorton) [inline src/include/msvc.h:115]: We have
```
#ifndef _M_AMD64
#error "Only x64 is supported with MSVC"
#endif
```
up on line 10, but it'd be useful to have comments in `WT_READ_BARRIER` and `WT_WRITE_BARRIER` that say they're x86 specific.

File: src/include/gcc.h [hunk 2]
```diff
@@ -282,8 +259,6 @@ WT_ATOMIC_FUNC(size, size_t, size_t *vp, size_t v)
         __asm__ volatile("bcr 15,0\n" ::: "memory"); \
     } while (0)
 #define WT_READ_BARRIER() WT_FULL_BARRIER()
-/* We only need a compiler barrier for s390x as its memory ordering is strong enough. */
```
Comment (ajmorton) [inline src/include/gcc.h:285]: This is weird. The `WEAK_MEMORDER` comment indicates that `WT_READ_BARRIER` should be a compiler barrier, but we actually use a full barrier? It's the same for i386.
Not a blocker for this PR but it warrants a ticket outside the scope of PM-3221

File: src/include/gcc.h [hunk 3]
```diff
@@ -164,16 +164,8 @@ WT_ATOMIC_FUNC(size, size_t, size_t *vp, size_t v)
...
-    do {                                         \
-        __asm__ volatile("lfence" ::: "memory"); \
-    } while (0)
-/* We only need a compiler barrier for x86 as its memory ordering is strong enough. */
-#define WT_READ_BARRIER_WEAK_MEMORDER() WT_BARRIER()
-#define WT_WRITE_BARRIER()                       \
-    do {                                         \
-        __asm__ volatile("sfence" ::: "memory"); \
-    } while (0)
+#define WT_READ_BARRIER() WT_BARRIER()
```
Comment (ajmorton) [inline src/include/gcc.h:167]:
```suggestion
/* We only need compiler barriers on x86 as Total Store Ordering (TSO) guarantees StoreStore and LoadLoad ordering. */
#define WT_READ_BARRIER() WT_BARRIER()
```

**Review by ajmorton — APPROVED** [review]
LGTM pending some comments and the `s_all` fix

File: src/include/gcc.h (hunk 2 above)
Comment (luke-pearson) [reply src/include/gcc.h:285]: I can create a follow up ticket for this. I don't think its a big deal as its not one of our main architectures.

Comment (luke-pearson) [reply src/include/gcc.h:285]: WT-12201

File: src/include/gcc.h (hunk 3 above)
Comment (luke-pearson) [reply src/include/gcc.h:168]: We'll need to change this to release / acquire when we get that.

File: src/include/msvc.h (hunk 1 above)
Comment (luke-pearson) [reply src/include/msvc.h:116]: [244c742](https://github.com/wiredtiger/wiredtiger/pull/10028/commits/244c742f639e500284b01cde5ac377f6ade3901f)

**Review by ajmorton — APPROVED** [review]

**Review by jeremythorp — APPROVED** [review]
LGTM!

**Comment by luke-pearson** [comment]
evergreen merge

**Pull request merged**